arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,315 @@
+    vil_image_view< pix_t > joint_nms_edges;
+
+    vil_math_image_sum( vil_plane( aligned_edges, 0 ),
+                        vil_plane( aligned_edges, 1 ),
+                        joint_nms_edges );
+
+    unsigned half_step = smoothing_half_step;
+    unsigned min_dim = std::min( joint_nms_edges.ni(), joint_nms_edges.nj() );
+
+    if( 2 * half_step + 1 >= min_dim )
+    {
+      half_step = ( min_dim - 1 ) / 2;
+    }
+
+    if( half_step != 0 )
+    {
+      vil_gauss_filter_2d( joint_nms_edges, combined_edges, smoothing_sigma,
+                           half_step );
+    }
+    else
+    {
+      vil_copy_reformat( joint_nms_edges, combined_edges );
+    }
+    vil_image_view< pix_t > all_channels( joint_nms_edges.ni(),
+                                          joint_nms_edges.nj(), 3 );
+    vil_plane( all_channels, 0 ).deep_copy( vil_plane( aligned_edges, 0 ) );
+    vil_plane( all_channels, 1 ).deep_copy( vil_plane( aligned_edges, 1 ) );
+    vil_plane( all_channels, 2 ).deep_copy( joint_nms_edges );
+    return all_channels;
+  }
+  return aligned_edges;
+}
+
+// ----------------------------------------------------------------------------
+aligned_edge_detection
+::aligned_edge_detection()
+  : d{ new priv{ this } }
+{
+  attach_logger( "arrows.vxl.aligned_edge_detection" );
+}
+
+// ----------------------------------------------------------------------------
+aligned_edge_detection
+::~aligned_edge_detection()
+{
+}
+
+// ----------------------------------------------------------------------------
+vital::config_block_sptr
+aligned_edge_detection
+::get_configuration() const
+{
+  // Get base config from base class
+  vital::config_block_sptr config = algorithm::get_configuration();
+
+  config->set_value( "threshold",
+                     d->threshold,
+                     "Minimum edge magnitude required to report as an edge "
+                     "in any output image." );
+  config->set_value( "produce_joint_output",
+                     d->produce_joint_output,
+                     "Set to false if we do not want to spend time computing "
+                     "joint edge images comprised of both horizontal and "
+                     "vertical information." );
+  config->set_value( "smoothing_sigma",
+                     d->smoothing_sigma,
+                     "Smoothing sigma for the output NMS edge density map." );
+  config->set_value( "smoothing_half_step",
+                     d->smoothing_half_step,
+                     "Smoothing half step for the output NMS edge density "
+                     "map." );
+
+  return config;
+}
+
+// ----------------------------------------------------------------------------
+void
+aligned_edge_detection
+::set_configuration( vital::config_block_sptr in_config )
+{
+  // Starting with our generated vital::config_block to ensure that assumed
+  // values are present. An alternative is to check for key presence before
```

Comment:
Because my inner grammarian is offended :wink::
```suggestion
  // Start with our generated vital::config_block to ensure that assumed values
  // are present. An alternative would be to check for key presence before
```